sentry/sentry.go

```diff
@@ -5,123 +5,48 @@ import (
 
 	"github.com/fabric8-services/fabric8-cluster/token"
-	"github.com/fabric8-services/fabric8-common/log"
+	"github.com/fabric8-services/fabric8-common/sentry"
 
 	"github.com/getsentry/raven-go"
-	goajwt "github.com/goadesign/goa/middleware/security/jwt"
+	"github.com/goadesign/goa/middleware/security/jwt"
 )
 
-// client encapsulates client to Sentry service
-// also has mutex which controls access to the client
-type client struct {
-	c       *raven.Client
-	sendErr chan func()
+type configuration interface {
+	GetSentryDSN() string
+	GetEnvironment() string
 }
 
-var (
-	sentryClient *client
-)
-
-// Sentry returns client declared inside package
-func Sentry() *client {
-	return sentryClient
-}
-
-// InitializeSentryClient initializes sentry client. This function returns
-// function that can be used to close the sentry client and error.
-func InitializeSentryClient(sentryDSN string, options ...func(*client)) (func(), error) {
-	c, err := raven.New(sentryDSN)
-	if err != nil {
-		return nil, err
-	}
-	sentryClient = &client{
-		c:       c,
-		sendErr: make(chan func()),
-	}
-	// set all options passed by user
-	for _, opt := range options {
-		opt(sentryClient)
-	}
-
-	// wait on errors to be sent on channel of client object
-	go sentryClient.loop()
-	return func() {
-		close(sentryClient.sendErr)
-	}, nil
-}
-
-// WithRelease helps you set release/commit of currently running
-// code while initializing sentry client using function InitializeSentryClient
-func WithRelease(release string) func(*client) {
-	return func(c *client) {
-		c.c.SetRelease(release)
-	}
-}
-
-// WithEnvironment helps you set environment the deployed code is
-// running in while initializing sentry client using function
-// InitializeSentryClient
-func WithEnvironment(env string) func(*client) {
-	return func(c *client) {
-		c.c.SetEnvironment(env)
-	}
-}
+// Initialize initializes sentry client
+func Initialize(config configuration, commit string) (func(), error) {
+	sentryDSN := config.GetSentryDSN()
 
-// waits on functions to be sent on channel
-// which are then executed
-func (c *client) loop() {
-	for op := range c.sendErr {
-		op()
-	}
+	return sentry.InitializeSentryClient(
+		&sentryDSN,
+		sentry.WithRelease(commit),
+		sentry.WithEnvironment(config.GetEnvironment()),
+		sentry.WithUser(extractUserInfo()))
 }
 
-// CaptureError sends error 'err' to Sentry, meanwhile also sets user
-// information by extracting user information from the context provided
-func (c *client) CaptureError(ctx context.Context, err error) {
-	// if method called during test which has uninitialized client
-	if c == nil {
-		return
-	}
-	// Extract user information. Ignoring error here but then before using the
-	// object user make sure to check if it wasn't nil.
-	user, _ := extractUserInfo(ctx)
-	reqID := log.ExtractRequestID(ctx)
-
-	c.sendErr <- func() {
-		if user != nil {
-			c.c.SetUserContext(user)
+func extractUserInfo() func(ctx context.Context) (*raven.User, error) {
```

Comment:
can't this `extractUserInfo()` can be part of f8-common.  I guess, this will be copied in future for each caller of `sentry.InitializeSentryClient()`. Also impl of `extractUserInfo()` looks generic and based on jwt and token stuff which is also at base for our services.

Reply:
We can't move it to f8-common as it is for now. The token manager won't be available in the context if extracted by common (because of different key). But we will see if we can improve it as part of https://github.com/fabric8-services/fabric8-cluster/issues/6